modules/currency.js

```diff
@@ -0,0 +1,206 @@
+    }
+    initCurrency(url);
+  } else {
+    // currency support is disabled, setting defaults
+    utils.logInfo('disabling currency support');
+    resetCurrency();
+  }
+}
+
+function initCurrency(url) {
+  utils.logInfo('Invoking initCurrency', arguments);
+
+  conversionCache = {};
+  currencySupportEnabled = true;
+
+  if(!oldBidResponse) {
+    utils.logInfo('Installing addBidResponse decorator for currency module', arguments);
+
+    oldBidResponse = bidmanager.addBidResponse;
+    bidmanager.addBidResponse = addBidResponseDecorator(bidmanager.addBidResponse);
+  }
+
+  ajax(url, function(response) {
+    try {
+      currencyRates = JSON.parse(response);
+      utils.logInfo('currencyRates set to ' + JSON.stringify(currencyRates));
+    } catch (e) {
+      utils.logError('failed to parse currencyRates response: ' + response);
+    }
+    currencyRatesLoaded = true;
+    processBidResponseQueue();
+  });
+}
+
+function resetCurrency() {
+  if(oldBidResponse) {
+    utils.logInfo('Uninstalling addBidResponse decorator for currency module', arguments);
+
+    bidmanager.addBidResponse = oldBidResponse;
+    oldBidResponse = undefined;
+  }
+
+  adServerCurrency = 'USD';
+  conversionCache = {};
+  currencySupportEnabled = false;
+  currencyRatesLoaded = false;
+  currencyRates = {};
+}
+
+export function addBidResponseDecorator(fn) {
+  return function(adUnitCode, bid) {
+    utils.logInfo('Invoking addBidResponseDecorator function', arguments);
+
+    // default to USD if currency not set
+    if(bid && !bid.currency) {
+      utils.logWarn('Currency not specified on bid.  Defaulted to "USD"');
+      bid.currency = 'USD';
+    }
+
+    if(bid) {
+      // execute immediately if the bid is already in the desired currency
+      if(bid.currency === adServerCurrency) {
+        return fn.apply(this, arguments);
+      }
+
+      bidResponseQueue.push(wrapFunction(fn, this, arguments));
+      if (!currencySupportEnabled || currencyRatesLoaded) {
+        processBidResponseQueue();
+      }
+    }
+  }
+}
+
+function processBidResponseQueue() {
+  utils.logInfo('Invoking processBidResponseQueue', arguments);
+
+  while (bidResponseQueue.length > 0) {
+    (bidResponseQueue.shift())();
+  }
+}
+
+function wrapFunction(fn, context, params) {
+  return function() {
+    utils.logInfo('Invoking wrapped function', arguments);
+    var bid = params[1];
+    if (bid !== undefined && 'currency' in bid && 'cpm' in bid) {
+      var fromCurrency = bid.currency;
+      try {
+        var conversion = getCurrencyConversion(fromCurrency);
+        bid.originalCpm = bid.cpm;
+        bid.originalCurrency = bid.currency;
+        if (conversion !== 1) {
+          bid.cpm = (parseFloat(bid.cpm) * conversion).toFixed(4);
+          bid.currency = adServerCurrency;
+        }
+      } catch (e) {
+        utils.logWarn('Returning NO_BID, getCurrencyConversion threw error: ', e);
+        params[1] = bidfactory.createBid(STATUS.NO_BID);
```

Comment:
this error condition will mess up the auction logic because it has a new `bid.id` not sure it's possible to fix though.